File: lib/engine/game.ts
import { GameState, BOARD_SIZE } from './types';
import { createEmptyBoard, canPlacePiece, placePiece, findCompletedLines, clearLines, hasValidPlacement } from './board';
import { getSmartPieces, getRandomPieces } from './pieces';
import { calculatePlacementScore, calculateClearScore } from './scoring';

function loadHighScore(): number {
  if (typeof window === 'undefined') return 0;
  try {
    const saved = localStorage.getItem('blockblast-highscore');
    return saved ? parseInt(saved, 10) : 0;
  } catch {
    return 0;
  }
}

function saveHighScore(score: number): void {
  if (typeof window === 'undefined') return;
  try {
    localStorage.setItem('blockblast-highscore', String(score));
  } catch {
    // silently fail if localStorage unavailable
  }
}

export function initGame(): GameState {
  const board = createEmptyBoard();
  const pieces = getRandomPieces(3);
  return {
    board,
    currentPieces: pieces,
    score: 0,
    combo: 0,
    streak: 0,
    isGameOver: false,
    turnNumber: 1,
    totalLinesCleared: 0,
    longestStreak: 0,
    highestCombo: 0,
    highScore: loadHighScore(),
    clearedThisTurn: false,
  };
}

function countPieceCells(shape: boolean[][]): number {
  let count = 0;
  for (const row of shape) {
    for (const cell of row) {
      if (cell) count++;
    }
  }
  return count;
}

export function handlePlacement(
  state: GameState,
  pieceIndex: number,
  row: number,
  col: number
): GameState | null {
  const piece = state.currentPieces[pieceIndex];
  if (!piece) return null;
  if (!canPlacePiece(state.board, piece, row, col)) return null;

  // Place the piece
  let newBoard = placePiece(state.board, piece, row, col);

  // Base points for placing the piece
  const cellsPlaced = countPieceCells(piece.shape);
  const placementPoints = calculatePlacementScore(cellsPlaced);

  // Check for completed lines
  const { rows, cols } = findCompletedLines(newBoard);
  const linesCleared = rows.length + cols.length;
  const cellsCleared = rows.length * BOARD_SIZE + cols.length * BOARD_SIZE
    - rows.length * cols.length; // subtract intersections

  // Clear lines
  if (linesCleared > 0) {
    newBoard = clearLines(newBoard, rows, cols);
  }

  // Calculate clear bonus (uses current combo for multiplier)
  const clearPoints = calculateClearScore(linesCleared, cellsCleared, state.combo);

  // Track if we cleared any line this turn
  const clearedThisTurn = state.clearedThisTurn || linesCleared > 0;

  const totalLinesCleared = state.totalLinesCleared + linesCleared;

  // Remove piece from tray
  const newPieces = [...state.currentPieces];
  newPieces[pieceIndex] = null;

  // Check if all 3 pieces have been placed — deal new set
  const allPlaced = newPieces.every(p => p === null);
  let finalPieces = newPieces;
  let newTurnNumber = state.turnNumber;
  let newCombo = state.combo;
  let newClearedThisTurn = clearedThisTurn;

  const totalPoints = placementPoints + clearPoints;
  const newScore = state.score + totalPoints;

  if (allPlaced) {
    // End of turn: update combo based on whether we cleared any line this turn
    if (clearedThisTurn) {
      // Cleared at least one line this turn — combo continues
      newCombo = state.combo + 1;
    } else {
      // No clears this turn — combo resets
      newCombo = 0;
    }

    finalPieces = getSmartPieces(3, newScore, newBoard);
    newTurnNumber = state.turnNumber + 1;
    newClearedThisTurn = false; // reset for new turn
  }

  const highestCombo = Math.max(state.highestCombo, newCombo);
  // Streak = same as combo for display purposes
  const streak = newCombo;
  const longestStreak = Math.max(state.longestStreak, streak);
  const highScore = Math.max(state.highScore, newScore);

  if (highScore > state.highScore) {
    saveHighScore(highScore);
  }

  const isGameOver = !hasValidPlacement(newBoard, finalPieces);

  return {
    board: newBoard,
    currentPieces: finalPieces,
    score: newScore,
    combo: newCombo,
    streak,
    isGameOver,
    turnNumber: newTurnNumber,
    totalLinesCleared,
    longestStreak,
    highestCombo,
    highScore,
    clearedThisTurn: newClearedThisTurn,
  };
}

File: training/env/game.py
import numpy as np
from .types import GameState, Piece, BOARD_SIZE
from .board import (
    create_empty_board,
    can_place_piece,
    place_piece,
    find_completed_lines,
    clear_lines,
    has_valid_placement,
)
from .pieces import get_random_pieces
from .scoring import calculate_placement_score, calculate_clear_score


def init_game(rng: np.random.Generator) -> GameState:
    board = create_empty_board()
    pieces = get_random_pieces(3, rng)
    return GameState(
        board=board,
        current_pieces=pieces,
        score=0,
        combo=0,
        placements_since_last_clear=0,
        is_game_over=False,
        turn_number=1,
        total_lines_cleared=0,
        highest_combo=0,
    )


def _count_piece_cells(shape: np.ndarray) -> int:
    return int(shape.sum())


def handle_placement(
    state: GameState,
    piece_index: int,
    row: int,
    col: int,
    rng: np.random.Generator,
) -> GameState | None:
    if piece_index < 0 or piece_index >= len(state.current_pieces):
        return None
    piece = state.current_pieces[piece_index]
    if piece is None:
        return None
    if not can_place_piece(state.board, piece, row, col):
        return None

    new_board = place_piece(state.board, piece, row, col)

    # Base points
    cells_placed = _count_piece_cells(piece.shape)
    placement_points = calculate_placement_score(cells_placed)

    # Line clears
    cleared_rows, cleared_cols = find_completed_lines(new_board)
    lines_cleared = len(cleared_rows) + len(cleared_cols)
    cells_cleared = (
        len(cleared_rows) * BOARD_SIZE
        + len(cleared_cols) * BOARD_SIZE
        - len(cleared_rows) * len(cleared_cols)
    )

    if lines_cleared > 0:
        new_board = clear_lines(new_board, cleared_rows, cleared_cols)

    # Combo logic
    new_combo = state.combo
    new_placements_since = state.placements_since_last_clear

    if lines_cleared > 0:
        new_combo = state.combo + 1
        new_placements_since = 0
    else:
        new_placements_since = state.placements_since_last_clear + 1
        if new_placements_since >= 3:
            new_combo = 0
            new_placements_since = 0

    # Clear score uses combo BEFORE incrementing
    clear_points = calculate_clear_score(lines_cleared, cells_cleared, state.combo)

    total_lines_cleared = state.total_lines_cleared + lines_cleared
    highest_combo = max(state.highest_combo, new_combo)

    # Remove piece from tray
    new_pieces = list(state.current_pieces)
    new_pieces[piece_index] = None

    # Deal new pieces if all 3 placed
    all_placed = all(p is None for p in new_pieces)
    new_turn_number = state.turn_number

    total_points = placement_points + clear_points
    new_score = state.score + total_points

    if all_placed:
        new_pieces = get_random_pieces(3, rng)
        new_turn_number = state.turn_number + 1

    is_game_over = not has_valid_placement(new_board, new_pieces)

    return GameState(
        board=new_board,
        current_pieces=new_pieces,
        score=new_score,
        combo=new_combo,
        placements_since_last_clear=new_placements_since,
        is_game_over=is_game_over,
        turn_number=new_turn_number,
        total_lines_cleared=total_lines_cleared,
        highest_combo=highest_combo,
    )
